### DISCUSSION TITLE
enabling vbox guest additions

### DISCUSSION THREAD

**Discussion by rjhornsby** [gusztavvargadr/packer#419] [Q&A] [answered]
Is there a generalized way to enable the installation of vbox guest additions? That is, I have a working custom build chained from win2019s - `ws2019s-rc-virtualbox-vagrant` if that helps explain what I've got.

The only thing I'm doing different from your repo is adding some packer stuff under `src/rc`, with the cookbooks, scripts etc that I need, and including (using `#load`) an entry to the root `build.cake` pointing to my `build.cake` stub:

```hcl
var ws2019s_rc = PackerTemplates_CreateWindows(
    "ws2019s-rc",
    "windows-server-2019-standard-rc",
    $"1809.0.{version}",
    ws2019s
);
```

I can see several examples where the variable `virtualbox_guest_additions_mode` is being set to `disabled`, but so far I can only find [one place](https://github.com/gusztavvargadr/packer/blob/master/src/w/packer/provisioners/chef/cookbooks/gusztavvargadr_packer_w/recipes/prepare.rb#L79) where it seems to be getting used. This usage is in the `w` source, so I think it's in the chain, but I'm not sure how to activate the logic to install the guest additions.

I thought it would be enough to tweak my `src/rc/template.json` like so:

```json
{
  "variables": {
    "name": "rc",
    "description": "install and configure base stack for RC development",
    "virtualbox_guest_additions_mode": "attach"
  }
}
```

but no luck so far. If I need to copy the logic from the cookbook in `src/w/packer...` I can do that, but with the variable set to `disabled` in so many places, it makes me think I'm missing some way to switch the behavior on.

**Comment by gusztavvargadr** [comment, marked as answer]
@rjhornsby thanks for reaching out.

There are a couple of different settings mentioned here. The packer setting `virtualbox_guest_additions_mode ` in the template JSONs is to control Packer's behavior regarding the guest additions. Since I install it using Chef, that is indeed set to `disabled` on purpose, so that the different approaches do not conflict. The `vbox?` condition you refer to has nothing to do with Packer's setting, it is one of [Chef's helpers for virtualization](https://docs.chef.io/infra_language/checking_hypervisors/).

Regarding the Chef provisioning, if you use the proper build chaining, the guest additions get installed during the core `Windows Server 2019` build, and there is no need to install them again in your extension - it is the whole point of the chaining, that you install the generic components and updates in one build, and your extensions in another one, resulting in shorter iterations usually.

If you don't use chaining, but build your image from scratch, you need to manually include the relevant recipes or copy the fragments to your own cookbook. Please take a look at how it is sovled for example in the IIS image by simply [including the generic Windows recipes](https://github.com/gusztavvargadr/packer/tree/060cc18634bc7a49dfdd82c4d1efd4fc16f2d235/src/iis/packer/provisioners/chef).

Another simple alternative can be to install the Guest Additions via [Chocolatey](https://community.chocolatey.org/packages/virtualbox-guest-additions-guest.install) as Chef handles these packages natively.

I hope this helps, please let me know.

**Comment by rjhornsby** [comment]
Thanks again for the insight and suggestions. I think I finally figured out that a big part of my problem was that I'd misunderstood or forgotten how the chaining actually worked. Then it was a matter of looking at the IIS example to sort of wrap my cookbooks in between `prepare`, `patch` and then finally `cleanup`.
